### PR TITLE
[DataLoader] Implementing communication processes for Map-style DataPipes

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68549

cc @SsnL @VitalyFedyunin @ejguan @NivekT

Differential Revision: [D32922676](https://our.internmc.facebook.com/intern/diff/D32922676)

### CODE REVIEW

File: test/test_dataloader.py [hunk 1]
```diff
@@ -2033,6 +2034,24 @@ def clean_me(process, req_queue, res_queue):
...
+        numbers_dp = SequenceWrapper(it)
+        (process, req_queue, res_queue, _thread_local_datapipe) = communication.eventloop.SpawnThreadForDataPipeline(numbers_dp)
+
+        process.start()
+        local_datapipe = communication.map.QueueWrapper(
+            communication.protocol.MapDataPipeQueueProtocolClient(req_queue, res_queue))
+        actual = list(local_datapipe)
+        clean_me(process, req_queue, res_queue)
+
+        self.assertEqual([(x, x) for x in range(100)], actual)
```
Comment (NivekT) [inline test/test_dataloader.py:2053]: Worth noting that currently `getitem` returns both `(key, value)`. Should we just return `value`?

File: torch/utils/data/communication/map.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            protocol.response_len(size)
+
+        elif isinstance(request, communication.messages.GetItemRequest):
+            while forever:
+                try:
+                    value = source_datapipe.nonblocking_getitem(request.key)
+                except NotAvailable:
+                    yield True
+                    continue
+                except IndexError as e:
```
Comment (NivekT) [inline torch/utils/data/communication/map.py:91]: I choose to trigger `StopIteration` when an invalid index is provided. This works well when the loop incrementally passes in indices and eventually goes out of bound. 

Do you think an alternative of raising IndexError and handling that is better?

File: torch/utils/data/communication/protocol.py [hunk 3]
```diff
@@ -62,80 +62,126 @@ def get_new_request(self, block=False):
...
-    def response_invalid(self):
+    def response_len(self, size):
         if not self.have_pending_request():
             raise Exception("Attempting to reply with pending request")
-        self.response_queue.put(communication.messages.InvalidStateResponse())
+        self.response_queue.put(communication.messages.LenResponse(size))
         self._req_received = None
 
-    def response_terminate(self):
+    def response_stop_iteration(self):
```
Comment (NivekT) [inline torch/utils/data/communication/protocol.py:90]: This handles the case when index goes out of bound.

File: torch/utils/data/dataloader_experimental.py [hunk 4]
```diff
@@ -55,6 +55,13 @@ def clean_me(thread, req_queue, res_queue):
             clean_me(thread, req_queue, res_queue)
 
 
+def sharding_worker_init_fn(worker_init_fn, worker_id):
```
Comment (NivekT) [inline torch/utils/data/dataloader_experimental.py:58]: I moved this inner function out to allow pickling.

File: torch/utils/data/communication/protocol.py (hunk 3 above)
Comment (VitalyFedyunin) [reply torch/utils/data/communication/protocol.py:90]: Please name it differently

File: torch/utils/data/dataloader_experimental.py [hunk 5]
```diff
@@ -33,7 +33,7 @@ def __iter__(self):
             for dp in self.datapipes:
                 if dp not in exclude_datapipes:
                     try:
-                        value = dp.nonblocking_next()
+                        value = dp.nonblocking_next()  # TODO: call nonblocking_getitem(index) for MapDataPipe
```
Comment (VitalyFedyunin) [inline torch/utils/data/dataloader_experimental.py:36]: Drop this comment

**Review by VitalyFedyunin — APPROVED** [review]
Overall looks good

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32922676).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32922676).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32922676).

**Pull request closed without merging**